### ISSUE TITLE
Object spread expressions are not contextually typed by their containing object literal

### ISSUE THREAD

**Issue by SaschaNaz** [Microsoft/TypeScript#15337]
**TypeScript Version:**  2.2.1

**Code**

```ts
// A *self-contained* demonstration of the problem follows...
interface Type {
    type: "type";
}

var type: Type;

type = { ... { type: "type" } }; // type 'string' is not assignable to type '"type"'

type = { ... { type: "type" as "type" } }; // no error
```

**Expected behavior:** I would expect that type broading not happen on object spread.

**Actual behavior:** Type broading happens on object spread.

**Comment by sandersn** [comment]
The problem is that contextual typing doesn't follow through a spread, so when `type: "type"` requests a contextual type, it asks for the contextual type of `{ type: "type" }`, which asks for the contextual type of `...{ type: "type" }`, which returns undefined.

In a normal assignment like `type = { type: "type" }`, the contextual type of `type = { type: "type" }` is `var type: Type`, which provides the literal type `"type"`.

This is definitely a bug, but it's pretty low priority because the only contextually typable things are literals, function expressions and object literals. The first two are not allowed (or non-sensical) in spreads and object literals can always be flattened into the parent object literal, which is also more efficient at runtime:

```ts
type = { ... { type: "type" } };
```

equivalent to

```ts
type = { type: "type" };
```

**Comment by SaschaNaz** [comment]
>object literals can always be flattened into the parent object literal,

Yes, if not doing this (probably the repro was too minimal):

```ts
interface SomeFetchParameterBag {
  param1: string;
  param2: string;
  param3: "A" | "B"
}

var injectParam3: Boolean;
myFetch({ param1, param2, ...injectParam3 && { param3 } });
```

**Comment by sandersn** [comment]
That repro still doesn't look complete. Here's one way (I also shortened the identifiers):

```ts
interface SomeFetchParameterBag {
  s: string;
  t: string;
  u: "A" | "B"
}
declare function myFetch(a: SomeFetchParameterBag): void;
function f(s: string, t: string, u: "A" | "B", inject: boolean) {
    myFetch({ s, t, ...inject && { u } });
}
```

But this doesn't fail because `u` already has a type that won't widen. If you use a literal instead of a parameter it does fail:

```ts
interface SomeFetchParameterBag {
  s: string;
  t: string;
  u: "A" | "B"
}
declare function myFetch(a: SomeFetchParameterBag): void;
function f(s: string, t: string, inject: boolean) {
    myFetch({ s, t, ...inject && { u: "A" } });
}
```

However, this is barely shorter than a top-level ternary:

```ts
    myFetch({ s, t, ...inject && { u: "A" } });
    myFetch({ s, t, u: inject ? "A" : undefined });
```

And I think the ternary is much less surprising on the first read.  Maybe if there are a lot of defaults like this, then it starts looking better. But in that case the default object should probably be declared on its own line.

**Comment by SaschaNaz** [comment]
>If you use a literal instead of a parameter it does fail:

Ah yes, I incorrectly skipped that part.

>And I think the ternary is much less surprising on the first read.

In some situation you won't want to define a property:

```js
new URLSearchParams({ u: undefined }).toString()
// "u=undefined"
```

**Comment by sandersn** [comment]
Fix is up at #15362
